lib/compat/getent-openbsd.c

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright (c) 2017 Balabit
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "compat/getent-openbsd.h"
+
+#ifdef __OpenBSD__
+#include <errno.h>
+
+int
+_compat_openbsd__getprotobynumber_r(int proto,
+                                    struct protoent *result_buf, char *buf,
+                                    size_t buflen, struct protoent **result)
+{
+  struct protoent_data *data_buf = (struct protoent_data *)buf;
+  int ret = 0;
+
+  if (buflen < sizeof(*data_buf))
+    ret = ERANGE;
+  else if (getprotobynumber_r(proto, result_buf, data_buf) == -1)
+    ret = errno;
```

Comment:
Usually in syslog-ng the curly bracket used with condition in order to improve readability (while there are exception from this, like early return cases). The other thing that I would ask you to do is exactly early returns.

My proposal would be the following:
```c
struct protoent_data *data_buf = (struct protoent_data *)buf;

if (buflen < sizeof(*data_buf)) 
   return ERANGE;

if (getprotobynumber_r(proto, result_buf, data_buf) == -1)
  return errno;
  
  *result = result_buf;
  return 0;
```

It has a behavior change as well, in case of an error `result` buffer is not touched, and that should be fine. The caller should not check the `result` buffer in case of an error as its content is undefined most of the time.
(This of course applies for the other places, just the comment is the same.)

Reply:
The Linux API requires that result be set to NULL on error and the code in syslog-ng which calls it relies on this behavior.  That is why I was careful to always clear it on failure.

Reply:
I pushed an update to this commit that adds braces and uses early return.